src/main/java/ch/njol/skript/SkriptAddon.java

```diff
@@ -21,123 +21,195 @@
+		return loadClasses(null, true, basePackage, true, subPackages);
+	}
+
+	@Nullable
+	private JarEntry @Nullable [] entryCache;
+
+	/**
+	 * This method resets the cache of jar entries used in {@link #loadClasses(Consumer, boolean, String, boolean, String...)}.
+	 * This method is meant for internal use, so you <i>probably</i> don't need it!
+	 * However, if you need loadClasses to load the same class multiple times, you <b>should</b> use this method.
+	 *
+	 * Note that this cache will be cleared when Skript stops accepting registrations.
+	 */
+	public void resetEntryCache() {
+		entryCache = null;
+	}
+
+	/**
+	 * Loads classes of the plugin by package. Useful for registering many syntax elements like Skript.
+	 *
+	 * Please note that if you need to load the same class multiple times,
+	 * you should call {@link #resetEntryCache()} each time you call this method.
+	 *
+	 * @param withClass A consumer that will run with each found class.
+	 * @param initialize Whether classes found in the package search should be initialized.
+	 * @param basePackage The base package to start searching in (e.g. 'ch.njol.skript').
+	 * @param recursive Whether to recursively search through the subpackages provided.
+	 * @param subPackages Specific subpackages to search in (e.g. 'conditions')
+	 *                    If no subpackages are provided, all subpackages of the base package will be searched.
 	 * @return This SkriptAddon
 	 */
-	public SkriptAddon loadClasses(String basePackage, String... subPackages) throws IOException {
-		assert subPackages != null;
-		JarFile jar = new JarFile(getFile());
+	@SuppressWarnings("ThrowableNotThrown")
+	public SkriptAddon loadClasses(@Nullable Consumer<Class<?>> withClass, boolean initialize, String basePackage, boolean recursive, String... subPackages) {
 		for (int i = 0; i < subPackages.length; i++)
 			subPackages[i] = subPackages[i].replace('.', '/') + "/";
 		basePackage = basePackage.replace('.', '/') + "/";
-		try {
-			List<String> classNames = new ArrayList<>();
 
-			for (JarEntry e : new EnumerationIterable<>(jar.entries())) {
-				if (e.getName().startsWith(basePackage) && e.getName().endsWith(".class")) {
+		int depth = !recursive ? StringUtils.count(basePackage, '/') + 1 : 0;
+
+		File file = getFile();
+		if (file == null) {
+			Skript.error("Unable to retrieve file from addon '" + getName() + "'. Classes will not be loaded.");
+			return this;
+		}
+
+		try (JarFile jar = new JarFile(file)) {
+			List<String> classNames = new ArrayList<>();
+			boolean hasWithClass = withClass != null;
+			if (entryCache == null)
+				entryCache = jar.stream().toArray(JarEntry[]::new);
+			for (int i = 0; i < entryCache.length; i++) {
+				JarEntry e = entryCache[i];
+				if (e == null) // This entry has already been loaded before
+					continue;
+				String name = e.getName();
+				if (name.startsWith(basePackage) && name.endsWith(".class") && (recursive || StringUtils.count(name, '/') <= depth)) {
```

Comment:
What if the subPackage is sth like `abc.def` (i.e. is a subpackage of a subpackage)?

Reply:
The "depth" tracking should account for subpackages - I've put in a fix :)